source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -0,0 +1,204 @@
+    insert_ = cache_.makeInsertContext(std::move(lookup_));
+    insert_->insertHeaders(headers, end_stream);
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (insert_) {
+    // TODO(toddmgreer) Wait for the cache if necessary.
+    insert_->insertBody(
+        data, [](bool) {}, end_stream);
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+void CacheFilter::onOkHeaders(Http::HeaderMapPtr&& headers,
+                              std::vector<AdjustedByteRange>&& response_ranges,
+                              uint64_t content_length, bool has_trailers) {
+  if (!lookup_) {
+    return;
+  }
+  response_has_trailers_ = has_trailers;
+  const bool end_stream = (content_length == 0 && !response_has_trailers_);
+  decoder_callbacks_->encodeHeaders(std::move(headers), end_stream);
+  if (end_stream) {
+    return;
+  }
+  if (content_length > 0) {
+    remaining_body_ = std::move(response_ranges);
+    // TODO(toddmgreer) handle multi-range requests.
+    ASSERT(remaining_body_.size() <= 1);
+    if (remaining_body_.empty()) {
+      remaining_body_.emplace_back(0, content_length - 1);
+    }
+    getBody();
+  } else {
+    lookup_->getTrailers([self = shared_from_this()](Http::HeaderMapPtr&& trailers) {
+      onTrailersAsync(self, std::move(trailers));
+    });
+  }
+}
+
+void CacheFilter::onUnusableHeaders() {
+  if (lookup_) {
+    decoder_callbacks_->continueDecoding();
+  }
+}
+
+void CacheFilter::onHeadersAsync(const CacheFilterSharedPtr& self, LookupResult&& result) {
+  switch (result.cache_entry_status) {
+  case CacheEntryStatus::RequiresValidation:
+  case CacheEntryStatus::FoundNotModified:
+  case CacheEntryStatus::UnsatisfiableRange:
+    ASSERT(false); // We don't yet return or support these codes.
+    // FALLTHROUGH
+  case CacheEntryStatus::Unusable: {
+    self->post([self] { self->onUnusableHeaders(); });
+    return;
+  }
+  case CacheEntryStatus::Ok:
+    self->post([self, headers = result.headers.release(),
+                response_ranges = std::move(result.response_ranges),
+                content_length = result.content_length,
+                has_trailers = result.has_trailers]() mutable {
+      self->onOkHeaders(absl::WrapUnique(headers), std::move(response_ranges), content_length,
+                        has_trailers);
+    });
+  }
+}
+
+void CacheFilter::getBody() {
+  ASSERT(!remaining_body_.empty());
+  CacheFilterSharedPtr self = shared_from_this();
+  lookup_->getBody(remaining_body_[0],
+                   [self](Buffer::InstancePtr&& body) { self->onBody(std::move(body)); });
+}
+
+void CacheFilter::onBodyAsync(const CacheFilterSharedPtr& self, Buffer::InstancePtr&& body) {
+  self->post([self, body = body.release()] { self->onBody(absl::WrapUnique(body)); });
+}
+
+void CacheFilter::onBody(Buffer::InstancePtr&& body) {
+  if (!lookup_) {
+    return;
+  }
+  ASSERT(!remaining_body_.empty());
```

Comment:
The body can't be empty?

Reply:
Clarified in assertion why there must be remaining body at this point.